File: sdks/typescript/src/legacy_canon.ts
/** `legacy-json-6dp` — the frozen canonicalization used by rr-trace/* and
 * draft-era reasoning-receipt/1 documents (spec §14). Mirrors
 * `protocol/legacy_canon.py` byte-for-byte, including Python `repr` float
 * formatting. */

import { ReceiptError } from "./errors.js";

export const LEGACY_CANONICALIZATION_ID = "legacy-json-6dp";

const encoder = new TextEncoder();

function round6(x: number): number {
  // Python: float(f"{x:.6f}") — format to 6dp then re-parse.
  // JS toFixed throws RangeError for |x| >= 1e21; every double in that
  // range is already integral so the value passes through unchanged.
  if (Math.abs(x) >= 1e21) return x;
  return Number(x.toFixed(6));
}

/** Python repr() of a float, emulated on JS numbers.
 * Rules: integral -> "N.0"; exponent form when exp10 < -4 or >= 16
 * (mantissa stripped of trailing zeros); otherwise decimal shortest repr
 * (which matches JS String(n) for the rounded values that occur here). */
function pyReprFloat(x: number): string {
  if (!Number.isFinite(x)) {
    throw new ReceiptError("malformed_number", "non-finite in legacy canon");
  }
  if (Object.is(x, -0)) return "-0.0";
  if (x === 0) return "0.0";
  if (Number.isInteger(x)) {
    const a = Math.abs(x);
    if (a >= 1e16) return expForm(x);
    return String(x) + ".0";
  }
  const a = Math.abs(x);
  const e = Math.floor(Math.log10(a));
  if (e < -4 || e >= 16) return expForm(x);
  return String(x);
}

function expForm(x: number): string {
  // Build "d.dddde+XX" like CPython repr for floats needing exponent form.
  const s = String(Math.abs(x));
  let digits: string;
  let exp: number;
  if (s.includes("e")) {
    // JS already in exponent form e.g. "1e+21", "1.5e-7"
    const m = /^(\d)(?:\.(\d+))?e([+-]\d+)$/.exec(s);
    if (!m) throw new ReceiptError("malformed_number");
    digits = (m[1] + (m[2] ?? "")).replace(/0+$/, "");
    exp = parseInt(m[3], 10);
  } else if (s.includes(".")) {
    // small values like "0.00001" -> exponent -5
    const m = /^0\.0*([1-9]\d*)$/.exec(s);
    if (m) {
      digits = m[1].replace(/0+$/, "");
      exp = -(s.indexOf(m[1]) - 1);
    } else {
      // e.g. "1.5" never reaches here (|x|>=1e16 is integer)
      digits = s.replace(".", "").replace(/0+$/, "");
      exp = s.indexOf(".") - 1;
    }
  } else {
    // large integer digits "10000000000000000"
    digits = s.replace(/0+$/, "");
    exp = s.length - 1;
  }
  const mant = digits.length === 1 ? digits : digits[0] + "." + digits.slice(1);
  const sign = x < 0 ? "-" : "";
  const expSign = exp < 0 ? "-" : "+";
  return `${sign}${mant}e${expSign}${String(Math.abs(exp)).padStart(2, "0")}`;
}

function norm(obj: unknown): unknown {
  if (typeof obj === "number") return round6(obj);
  if (Array.isArray(obj)) return obj.map(norm);
  if (obj !== null && typeof obj === "object") {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(obj)) out[k] = norm(v);
    return out;
  }
  return obj;
}

function encode(v: unknown): string {
  if (v === null) return "null";
  if (typeof v === "boolean") return v ? "true" : "false";
  if (typeof v === "number") return pyReprFloat(v);
  if (typeof v === "string") {
    // json.dumps(ensure_ascii=True): escape non-ASCII as \uXXXX (+surrogates).
    let out = '"';
    for (let i = 0; i < v.length; i++) {
      const cp = v.charCodeAt(i);
      switch (cp) {
        case 0x22: out += '\\"'; break;
        case 0x5c: out += "\\\\"; break;
        case 0x08: out += "\\b"; break;
        case 0x09: out += "\\t"; break;
        case 0x0a: out += "\\n"; break;
        case 0x0c: out += "\\f"; break;
        case 0x0d: out += "\\r"; break;
        default:
          if (cp < 0x20 || cp > 0x7e) {
            out += "\\u" + cp.toString(16).padStart(4, "0");
          } else {
            out += v[i];
          }
      }
    }
    return out + '"';
  }
  if (Array.isArray(v)) return "[" + v.map(encode).join(",") + "]";
  if (typeof v === "object") {
    // json.dumps(sort_keys=True) sorts by code point; UTF-8 byte order is
    // equivalent to code-point order and differs from JS UTF-16 ordering
    // for astral keys.
    const keys = Object.keys(v as object)
      .map((k) => ({ k, b: encoder.encode(k) }))
      .sort((a, b) => {
        const m = Math.min(a.b.length, b.b.length);
        for (let i = 0; i < m; i++) {
          if (a.b[i] !== b.b[i]) return a.b[i] - b.b[i];
        }
        return a.b.length - b.b.length;
      })
      .map(({ k }) => k);
    const parts = keys.map(
      (k) => encode(k) + ":" + encode((v as Record<string, unknown>)[k]),
    );
    return "{" + parts.join(",") + "}";
  }
  throw new ReceiptError("malformed_document");
}

export function legacyCanonicalBytes(doc: Record<string, unknown>): Uint8Array {
  return encoder.encode(encode(norm(doc)));
}

File: sdks/rust/src/legacy_canon.rs
//! `legacy-json-6dp` — the frozen canonicalization used by `rr-trace/*` and
//! draft-era `reasoning-receipt/1` documents (spec §14). Mirrors
//! `protocol/legacy_canon.py` byte-for-byte, including CPython `repr` float
//! formatting.

use serde_json::Value;

pub const LEGACY_CANONICALIZATION_ID: &str = "legacy-json-6dp";

fn round6(x: f64) -> f64 {
    if !x.is_finite() {
        return x;
    }
    if x.abs() >= 1e21 {
        return x; // Python f"{x:.6f}" handles big values; every f64 there is integral
    }
    format!("{x:.6}").parse().unwrap_or(x)
}

/// CPython repr() of a float.
fn py_repr_float(x: f64) -> String {
    if x == 0.0 {
        return if x.is_sign_negative() { "-0.0".into() } else { "0.0".into() };
    }
    if x == x.trunc() && x.abs() < 1e16 {
        return format!("{}.0", x as i64);
    }
    let e = x.abs().log10().floor() as i32;
    if e < -4 || e >= 16 || (x == x.trunc() && x.abs() >= 1e16) {
        return exp_form(x);
    }
    // Shortest round-trip repr: Rust's {} on f64 gives the same digits.
    format!("{x}")
}

fn exp_form(x: f64) -> String {
    // Produce "d.ddde+XX" like CPython repr.
    let s = format!("{:e}", x.abs()); // e.g. "1.5e16", "1e-5", "1.234e20"
    let (mant, exp) = s.split_once('e').unwrap();
    let mut digits: String = mant.chars().filter(|c| *c != '.').collect();
    while digits.len() > 1 && digits.ends_with('0') {
        digits.pop();
    }
    let exp: i32 = exp.parse().unwrap();
    let mantissa = if digits.len() == 1 {
        digits
    } else {
        format!("{}.{}", &digits[..1], &digits[1..])
    };
    format!(
        "{}{}e{}{:02}",
        if x < 0.0 { "-" } else { "" },
        mantissa,
        if exp < 0 { "-" } else { "+" },
        exp.abs()
    )
}

fn norm(v: &Value) -> Value {
    match v {
        Value::Number(n) => {
            if let Some(f) = n.as_f64() {
                if !f.is_finite() {
                    return v.clone();
                }
                if n.is_f64() {
                    // Preserve floatness: integral floats stay floats after round6
                    let r = round6(f);
                    if r == r.trunc() && r.abs() < 1e16 {
                        return Value::from(r); // still f64 → repr "N.0"
                    }
                    Value::from(r)
                } else {
                    v.clone() // ints pass through unchanged
                }
            } else {
                v.clone()
            }
        }
        Value::Array(a) => Value::Array(a.iter().map(norm).collect()),
        Value::Object(m) => Value::Object(m.iter().map(|(k, v)| (k.clone(), norm(v))).collect()),
        _ => v.clone(),
    }
}

fn encode(v: &Value, out: &mut String) {
    match v {
        Value::Null => out.push_str("null"),
        Value::Bool(b) => out.push_str(if *b { "true" } else { "false" }),
        Value::Number(n) => {
            if let Some(f) = n.as_f64() {
                if n.is_f64() {
                    out.push_str(&py_repr_float(f));
                } else {
                    out.push_str(&n.to_string());
                }
            }
        }
        Value::String(s) => encode_str(s, out),
        Value::Array(a) => {
            out.push('[');
            for (i, x) in a.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                encode(x, out);
            }
            out.push(']');
        }
        Value::Object(m) => {
            // json.dumps(sort_keys=True): Python str ordering = code-point
            // order = UTF-8 byte order.
            let mut keys: Vec<&String> = m.keys().collect();
            keys.sort_by(|a, b| a.as_bytes().cmp(b.as_bytes()));
            out.push('{');
            for (i, k) in keys.iter().enumerate() {
                if i > 0 {
                    out.push(',');
                }
                encode_str(k, out);
                out.push(':');
                encode(&m[*k], out);
            }
            out.push('}');
        }
    }
}

/// json.dumps(ensure_ascii=True): escape non-ASCII as \uXXXX (+ surrogate pairs).
fn encode_str(s: &str, out: &mut String) {
    out.push('"');
    for c in s.chars() {
        match c {
            '"' => out.push_str("\\\""),
            '\\' => out.push_str("\\\\"),
            '\u{08}' => out.push_str("\\b"),
            '\t' => out.push_str("\\t"),
            '\n' => out.push_str("\\n"),
            '\u{0c}' => out.push_str("\\f"),
            '\r' => out.push_str("\\r"),
            c if (c as u32) < 0x20 => out.push_str(&format!("\\u{:04x}", c as u32)),
            c if (c as u32) > 0x7e => {
                let cp = c as u32;
                if cp > 0xffff {
                    let v = cp - 0x10000;
                    out.push_str(&format!(
                        "\\u{:04x}\\u{:04x}",
                        0xd800 + (v >> 10),
                        0xdc00 + (v & 0x3ff)
                    ));
                } else {
                    out.push_str(&format!("\\u{:04x}", cp));
                }
            }
            c => out.push(c),
        }
    }
    out.push('"');
}

pub fn legacy_canonical_bytes(doc: &Value) -> Vec<u8> {
    let mut out = String::new();
    encode(&norm(doc), &mut out);
    out.into_bytes()
}
